src/solvers/td3.ml

```diff
@@ -234,10 +244,45 @@ module WP =
         )
       and eq x get set =
         if tracing then trace "sol2" "eq %a\n" S.Var.pretty_trace x;
-        eval_rhs_event x;
+        
         match S.system x with
         | None -> S.Dom.bot ()
-        | Some f -> f get set
+        | Some f ->
+          if skip_unchanged_rhs then
+            let all_deps_unchanged =
+              match HM.find_option dep_vals x with
+              | None -> None
+              | Some (oldv, deps) ->
+                let deps_inorder = List.rev deps in
+                if List.for_all (fun (var, value) -> S.Dom.equal (get var) value) deps_inorder then
```

Comment:
Intuitively this reversal makes sense to have the dependencies in the same order as they were added, but is it really necessary? Because the only operation on `deps_inorder` is a `List.for_all` for which it doesn't really matter what the order of elements is.

Because this reversal copies the list each time and immediately discards it after `List.for_all`, creating garbage. Or if it is really necessary/useful for some reason, maybe do the reversal when the dependencies are connected. That way it's done once per unchanged unknown, rather than as many times as the unknown is later queried again.

Reply:
I think in principle, the same unknown can be queried in an RHS twice, and in the presence of side-effects  (e.g. from other unknowns that are solved), I think there is no guarantee that both times the same value is returned. If this is the case, it might make a difference because the first read has read an outdated value while later ones have read the current one.
This means that storing as a map (Var -> Val) is definitely out, and I am a bit worried that having the dependencies in another order might also be problematic, though I do not have an example yet.

Reply:
If the same queried unknown changed in between due to a side effect, then it cannot be reused at all, because later when doing these checks `get var` will return the same value in both cases (because between these there are no side effects), so one of the two equality checks must necessarily fail.

Reply:
> because between these there are no side effects

Is this actually guaranteed though? What if one of the `get`s in turn triggers an evaluation (including side-effects) of some other unknown?

Reply:
The other point that I forgot to mention is that if I do it in the order that the unknowns were actually queried, I will not issue calls to get (and thus potentially to `solve`) for unknowns that then won't be used in the RHS.

Consider the worst case where I depend on many variables and do this comparison without reversing the list. The changed value for the first unknown the RHS asks for might mean that none of the other unknowns queried previously are consulted anymore. If I reverse the list, I will stop immediately at the first changed unknown, having only asked for things that the RHS will ask for as well. If I don't, I will consider all others to then notice that the endeavour was pointless, having queried things that the RHS will not even ask for anymore.

Reply:
> Is this actually guaranteed though? What if one of the `get`s in turn triggers an evaluation (including side-effects) of some other unknown?

Suppose one call of `system x` does:
1. `get y` and gets `d_1` (and caches `(y, d_1)`),
2. something else which destabilizes `y`,
3. `get y` again and gets `d_2` (and caches `(y, d_2)`).

After this, `rho y = d_2`.

The following call of `system x` would first check if `get y = d_1`, which wouldn't hold (something might've increased `rho y` further, but it wouldn't go back down to `d_1`). If it somehow did, then the second `system x` evaluation would be causing the exact same increasing transition from `d_1` to `d_2` for `y` again, which seems to violate TD's termination.
So it's not even about what happens between the `get`s of the subsequent call.

Side-effected variables (leaves in TD3 paper terminology) only join and widen contributions, but there's no narrowing for leaves. If there was, then I suppose it might be possible for a leaf to narrow back down to `d_1` again, but otherwise I cannot see how it could happen.

> Consider the worst case where I depend on many variables and do this comparison without reversing the list. The changed value for the first unknown the RHS asks for might mean that none of the other unknowns queried previously are consulted anymore. If I reverse the list, I will stop immediately at the first changed unknown, having only asked for things that the RHS will ask for as well. If I don't, I will consider all others to then notice that the endeavour was pointless, having queried things that the RHS will not even ask for anymore.

That's a good point. In my versions with exception aborting this difference was avoided by only ever doing `get`s that RHS does, so it always matches the order that the RHS queries them in and doesn't query anything additional accidentally.

Intuitively it seems to me that if there are unknowns that are no longer queried, then `get`ting them wouldn't do real work either, but just look them up. If something else changed them, then they are still in use for something else, meaning that even if that `get` triggers a real `solve`, it would solve something that would've been needed to eventually solved anyway.

Continuing from that last point, if such things somehow happen, then if the preliminary `get` for comparison does an actual `solve` underneath, then that could drastically change what is solved in what order. In particular, this could change the order of some side effects, where in turn widening might cause different results because of it (since `widen` is not commutative).
So it's far from just observing the solving, but the preliminary checks do actually change the solving structure.

Of course my entire point about `List.rev` was just a guess. Profiling with perf and memtrace should directly show if that's anything significant at all.